Add new AD derivative type

lindsayad opened the pull request:
Switch from the `NumberArray` derivative type to the new derivative
type `SemiDynamicSparseNumberArray`. This template class has
data members to hold both the derivative values and the derivative
indices, meaning that we can take greatest advantage of sparsity.
The backing storage type is `std::array` so that storage is static
and we don't see any slow down from dynamic memory allocation.
We avoid looping over the whole size of the std::array by
maintaining a dynamic size data member. The only real draw-back
to this derivative type is that we'll hit an assertion if our sparse
size ever exceeds the static storage size of the std::array.

Also update libmesh. Summary of changes:
 - Update MetaPhysicL
  - Reduce header dependencies
  - Disabled glibc++ debug defines by default.
  - Setting the right MPI communicator for VTK IO.
  - More digits for SEVENTH-order quadrature on TRIs.
  - Install metis.h
  - MeshBase::add_elem/node_datum APIs
  - Implement MONOMIAL_VEC
  - Slepc $PETSC_ARCH fix
  - Intro/ex2 options + random typo fix
  - Make mallocs uniformly an error in PetscMatrix
  - Boost multiprecision
  - Don't try remote_elem->set_parent()
  - Unit test updates
  - Bugfix in rb_data_deserialization.C
  - Real literals
  - libmesh_cppunit.h refactor
  - Why are you always failing?
  - Add jacobian_tolerance on DirichletBoundary class.

lindsayad committed:
Switch to SDSNA

Switch from the `NumberArray` derivative type to the new derivative
type `SemiDynamicSparseNumberArray`. This template class has
data members to hold both the derivative values and the derivative
indices, meaning that we can take greatest advantage of sparsity.
The backing storage type is `std::array` so that storage is static
and we don't see any slow down from dynamic memory allocation.
We avoid looping over the whole size of the std::array by
maintaining a dynamic size data member. The only real draw-back
to this derivative type is that we'll hit an assertion if our sparse
size ever exceeds the static storage size of the std::array. You can get
around this by increasing the static size of the array, but you will
incur some (small?) memory overhead because of the material properties
and variable values (dynamic storage containers std::vector and MooseArray
respectively) that hold MOOSE's DualReal objects.

lindsayad committed:
Non-linear algorithm documentation

Touches on:

- Newton's method
- Krylov methods
- Jacobian construction
- Automatic differentiation

lindsayad committed:
Update libMesh

  - ExodusII: Read and write sideset data
  - Include libmesh_logging.h to define LOG_SCOPE
  - Add "RATIONAL_BERNSTEIN" to string_to_enum maps
  - Allow TRI3SUBDIVISION elements in more quadrature rules.
  - Rational basis functions
  - Fix FE unit tests and the broken FE they caught
  - Fixup MetaPhysicL update
  - Add missing array header include
  - Fix missing includes in --disable-optional builds
  - Update MetaPhysicL
  - Reduce header dependencies
  - Disabled glibc++ debug defines by default.
  - Setting the right MPI communicator for VTK IO.
  - More digits for SEVENTH-order quadrature on TRIs.
  - Install metis.h
  - MeshBase::add_elem/node_datum APIs
  - Implement MONOMIAL_VEC
  - Slepc $PETSC_ARCH fix
  - Intro/ex2 options + random typo fix
  - Make mallocs uniformly an error in PetscMatrix
  - Boost multiprecision
  - Don't try remote_elem->set_parent()
  - Unit test updates
  - Bugfix in rb_data_deserialization.C
  - Real literals
  - libmesh_cppunit.h refactor
  - Why are you always failing?
  - Add jacobian_tolerance on DirichletBoundary class.

Refs #000

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/387727/) on f27b28a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13963/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13963/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3a8848c6f90387a0aa62d7dc35fa4c6ad2ea4c93`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit da98c0178b4d03f222d6b02c1a701eea8a38af5e
+Subproject commit e8ea542453e745f2492d84654fb4a564a47d0491
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

lindsayad commented:
@tophmatthews I believe you've said you've been unable to run some 3D problems with AD. Is that right? For those 3D problems are you using a second order mesh?

tophmatthews commented:
I can't remember...but probably 2nd order due to mechanics

lindsayad committed:
No longer need AD_MAX_DOFS_PER_ELEM check

lindsayad committed:
Missing include apparently

lindsayad committed:
Bump AD_MAX_DOFS_PER_ELEM to 81

81 seems like a logical choice because if one had a second order
mesh and a vector-like variable that had coupling between all its
components, then one would need a derivative size of 81. This choice
should be discussed with the rest of the @idaholab/moose-team.
Bumping should not cause any cpu performance penalty because we only
loop over dof-index coupling that we've explictly computed through
sparsity operations. However, there may be some memory penalty. I
figure @rwcarlsen and I can do some memory profiling to assess
this, most importantly on simulations that use AD and stateful
material properties.

lindsayad commented:
@dschwen and @tophmatthews Do you guys have an input file that uses both AD and stateful material properties? This would be a good candidate for memory profiling

tophmatthews commented:
I'm just starting to use AD so nothing solid, I'll try and find one for you.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/387798/) on fae2fca wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13963/site)

This comment will be updated on new commits.

tophmatthews commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1330,11 +1330,6 @@ MooseVariableData<OutputType>::computeAD(const unsigned int num_dofs, const unsi
 
-  // Hopefully this problem can go away at some point
-  if (ad_offset + num_dofs > AD_MAX_DOFS_PER_ELEM)
-    mooseError("Current number of dofs per element is greater than AD_MAX_DOFS_PER_ELEM of ",
```
Why is this? even if it's in 3D with a bunch of variables? 81 will always be the most?

tophmatthews replied:
thats fine, as long as it is an obvious error so its easy to debug if 81 is not enough.

lindsayad commented:
No it will not always be the most. One could conjure up many scenarios in which a bigger vector is needed, but the question here is what makes sense as a default. I would argue that a MOOSE configure may make more and more sense as people do more “intense” simulations with AD in MOOSE (eg 3D tight coupling). A MOOSE configure would allow setting the derivative vector size for a given application. Hell a MOOSE configure could even select a derivative type. 

> On Aug 23, 2019, at 4:32 PM, Topher Matthews <notifications@github.com> wrote:
> 
> @tophmatthews commented on this pull request.
> 
> In framework/src/variables/MooseVariableData.C:
> 
> > @@ -1330,11 +1330,6 @@ MooseVariableData<OutputType>::computeAD(const unsigned int num_dofs, const unsi
>    mooseAssert(_var.kind() == Moose::VarKindType::VAR_AUXILIARY || ad_offset || !_var_num,
>                "Either this is the zeroth variable or we should have an offset");
>  
> -  // Hopefully this problem can go away at some point
> -  if (ad_offset + num_dofs > AD_MAX_DOFS_PER_ELEM)
> -    mooseError("Current number of dofs per element is greater than AD_MAX_DOFS_PER_ELEM of ",
> Why is this? even if it's in 3D with a bunch of variables? 81 will always be the most?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

lindsayad committed:
DataIO for new derivative type

This also fixes a bug where I didn't have a corresponding
dataLoad function for loading an array of DualNumbers. I have
no idea how this worked before

lindsayad committed:
Do not use the .size() idiom

This doesn't work because we expect the index by the actual
dof number, NOT by its element location in the underlying
data array.

lindsayad committed:
Add AD configuration options

lindsayad committed:
Update DualReal for being configurable

lindsayad committed:
Change framework level derivative insertion

lindsayad committed:
Replace all metaphysicl header includes with DualRealOps

tophmatthews commented:
2X time hit compared to non-AD, or AD not using this PR @rwcarlsen ?

lindsayad commented:
> 2X time hit compared to non-AD, or AD not using this PR @rwcarlsen ?

Old AD. I'm currently setting up our make system to compile by default with the old AD derivative type, but if the user specifies something like `export SPARSE=true` in their environment, then MOOSE will compile and use the sparse derivative type.

lindsayad committed:
Allow users to pull in both SparseOps and NumberArrayOps if they want

lindsayad committed:
Adapt modules to configurable DualReal

lindsayad committed:
Adapt unit tests to configureable DualReal

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/390364/) on f0405ae wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

lindsayad committed:
Work around strange minimum clang error

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/390485/) on becad95 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

lindsayad committed:
Add MOOSE configure

lindsayad committed:
Add auto-generated files

lindsayad committed:
requisite moose.mk changes

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit da98c0178b4d03f222d6b02c1a701eea8a38af5e
+Subproject commit b79a98038bec4ca58ecf374e39f17a012d5db9db
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/391456/) on 6547432 : invalidated by @permcody

WTH?

moosebuild commented:
Job [64bit dof id Test](https://civet.inl.gov/job/391604/) on 2fe956b : invalidated by @lindsayad

this is insane

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/391612/) on 2fe956b : invalidated by @lindsayad

MOOSE objects should not be compiling before the configure symlink is done per the rule in moose.mk...

permcody commented:
@idaholab/moose-team - Here is the PR with the Make issue that I was telling you guys about in the cube. Alex has made a few minor changes to moose.mk to force the symlinking of the MooseConfig.h file before anything else occurs (well almost...) As Alex said, we are getting lucky since we have a missing dependency between moose and header symlinking already. Alex said that the code coverage target is always failing with this change.

lindsayad commented:
I'm so bamboozled. If I log into the code coverage test box and run with the exact same configuration and commands, then I see exactly what I want to see (moose configure symlinking before (non-contrib) moose sources being built):

<pre>
Rebuilding symlinks in /opt/civet/build_0/moose/framework/build/header_symlinks
Copying default MOOSE configuration to: /opt/civet/build_0/moose/framework/include/base/MooseConfig.h...
Checking if header needs updating: /opt/civet/build_0/moose/framework/include/base/MooseRevision.h...
Creating Unity Directory /opt/civet/build_0/moose/framework/build/unity_src
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcrecpp.cc...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_stringpiece.cc...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_compile.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_byte_order.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_globals.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_ucd.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_chartables.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_config.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_fullinfo.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_xclass.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_refcount.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_ord2utf8.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_dfa_exec.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_jit_compile.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_study.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_newline.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_tables.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_string_utils.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_valid_utf8.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_version.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_exec.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_maketables.c...
Compiling C (in opt mode) /opt/civet/build_0/moose/framework/contrib/pcre/src/pcre_get.c...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/gtest/gtest-all.cc...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/hit/parse.cc...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/hit/lex.cc...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/hit/braceexpr.cc...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/create_file.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/exodiff.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/check.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/iqsort.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/node_set.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/Tolerance.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/ED_SystemInterface.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/exo_block.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/exo_entity.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/STRINGLIB_tokenize.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/FileInfo.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/add_to_log.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/SL_tokenize.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/util.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/map.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/GetLongOpt.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/stringx.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/exoII_read.C...
Compiling C++ (in opt mode) /opt/civet/build_0/moose/framework/contrib/exodiff/side_set.C...
<b>Symlinking MOOSE configure
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/src_Unity.C</b>
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/interfaces_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/distributions_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/loops_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/variables_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/interfacekernels_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/problems_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/preconditioners_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/actions_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/ics_Unity.C
Creating Unity /opt/civet/build_0/moose/framework/build/unity_src/auxkernels_Unity.C
</pre>

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/391612/) on 2fe956b : invalidated by @lindsayad

moosebuild commented on petsc:
```diff
@@ -1 +1 @@
-Subproject commit 8fb8dbaab76e3eb36061f8be821d8703451f6653
+Subproject commit 2c2c1ef19bc72ffdad11a06b56b17744e8134136
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/391859/) on 3cf1045 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit da98c0178b4d03f222d6b02c1a701eea8a38af5e
+Subproject commit ca543b0a6d9af487d4b7d1e64b890aa777a3bb9c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/392182/) on 8cdbe7d wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/392252/) on 29d8763 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

permcody commented on framework/include/utils/NumberArrayOps.h:
```diff
@@ -0,0 +1,22 @@
+inline void
+derivInsert(NumberArray<N, Real> & derivs, unsigned int index, Real value)
+{
+  derivs[index] = value;
```
mooseAssert?

lindsayad replied:
Added

permcody commented on framework/include/utils/SparseOps.h:
```diff
@@ -0,0 +1,24 @@
+            unsigned int index,
+            Real value)
+{
+  derivs.insert(index) = value;
```
mooseAssert?

lindsayad replied:
Added `try-catch` because we don't known *a priori* whether an insertion into the sparse container will cause problems. This can only be determined within the underlying `MetaPhysicL` machinery, where if something bad happens an exception is thrown.

permcody commented on framework/moose.mk:
```diff
@@ -72,7 +82,7 @@ $(1):
 endef
 
-include_files	:= $(shell find $(FRAMEWORK_DIR)/include -regex "[^\#~]*\.h")
+include_files	:= $(shell find $(FRAMEWORK_DIR)/include \( -regex "[^\#~]*\.h" ! -name "*MooseConfig.h" \))
```
Hopefully we don't run into weird platform differences in find. I really hate find.

permcody commented on framework/src/interfaces/OrientedBoxInterface.C:
```diff
@@ -13,7 +13,7 @@
 #include "MooseError.h"
 
-// libMesh incluces
+#include "libmesh/auto_ptr.h"
```
Not sure why this is here. Did you use it in a downstream object?

lindsayad replied:
There is a `libmesh_make_unique` call here in this file. That macro is defined in `auto_ptr.h`.

permcody commented on modules/functional_expansion_tools/src/series/CompositeSeriesBasisInterface.C:
```diff
@@ -10,6 +10,8 @@
 #include "CompositeSeriesBasisInterface.h"
 #include "Legendre.h"
 
+#include "libmesh/auto_ptr.h"
```
Again - it's unclear why this is here. I assume something downstream...

lindsayad replied:
`libmesh_make_unique`

permcody reviewed (commented):
This looks OK to me. It's a big change though so I wouldn't mind having others on the @idaholab/moose-team weigh in before this gets merged.

Actually the change really isn't that big. The fact that we are adding configure in this PR is the big part of the change.

lindsayad committed:
Prefix our defines

lindsayad committed:
Document configure

lindsayad committed:
Reconfigure

lindsayad committed:
Debug guards for derivative insertion

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/392562/) on 261dc39 : invalidated by @permcody

Not firing....

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/392595/) on 261dc39 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/13963/coverage)

tophmatthews commented:
@lindsayad This PR results in a reduced `MOOSE_AD_MAX_DOFS_PER_ELEM` correct? Before this I had to use 102 for four variables in 2DRz (disp_x, disp_y, temperature, lm_for_mortar), but now the default of 50 is adequate with this PR?

lindsayad commented:
The old derivative type is still the default in MOOSE. You can control the derivative type and the derivative vector length using MOOSE’s new configure. I’m away from my computer so I can’t tell you the exact option names right now, but you can see them if you cd into the framework directory and run ‘./configure —help’. I’ll also be sending out an announcement to the list tomorrow. 

> On Sep 9, 2019, at 3:00 PM, Topher Matthews <notifications@github.com> wrote:
> 
> @lindsayad This PR results in a reduced MOOSE_AD_MAX_DOFS_PER_ELEM correct? Before this I had to use 102 for four variables in 2DRz (disp_x, disp_y, temperature, lm_for_mortar), but now the default of 50 is adequate with this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tophmatthews commented:
Sure, I get that, but not touching anything in configure, which means it should use the same derivative type as before, results in less required `MOOSE_AD_MAX_DOFS_PER_ELEM` with the new this PR, with the exact same input file. Or at least, it doesn't error out with `MOOSE_AD_MAX_DOFS_PER_ELEM=50` with this PR, but does before this PR.

lindsayad commented:
> Sure, I get that, but not touching anything in configure, which means it should use the same derivative type as before, results in less required `MOOSE_AD_MAX_DOFS_PER_ELEM` with the new this PR, with the exact same input file. Or at least, it doesn't error out with `MOOSE_AD_MAX_DOFS_PER_ELEM=50` with this PR, but does before this PR.

I think this is probably because of this discussion [here](https://github.com/idaholab/moose/pull/13963/files#r317330535). Can you email me an input file or tell me what I need to reproduce this? (e.g. element type and mortar?) I'll be looking at your convergence issue today.

tophmatthews commented:
The input I sent you last week reproduces this. Thanks for taking a look

lindsayad commented:
Ok, yea it is because of [that change](https://github.com/idaholab/moose/pull/13963/files#r317330535). If you run in devel or dbg mode, you will receive this message:
```
Time Step 1, time = 100, dt = 100
Reiniting because of ghosting ...................................                          [ 36.54 s]
Computing initial residual ...                                                             [  4.53 s]
 0 Nonlinear |R| = 1.600143e+00


Assertion `index < MOOSE_AD_MAX_DOFS_PER_ELEM' failed
The requested derivative index 50 is not less than 50. You can run `configure --with-derivative-size=<n>` to request a larger derivative container.
at /Users/lindad/projects/moose/framework/build/header_symlinks/NumberArrayOps.h, line 26
```
Perhaps I will add back the error message in the `MooseVariableData::computeAD` method while keeping the generic assertion to catch other AD uses.

tophmatthews commented:
Ah....that seems obvious now....
